### PR TITLE
Allow use of URL-namespaced claims

### PR DESCRIPTION
**Pull request by willthames**
Private namespaced claims are recommended to be prefixed with URLs,
but having dots in domain names means that working out nested key
structure by splitting on dots doesn't work.
    
Instead we assume that if a claim starts with http:// or https://
then everything up to the third slash is part of the top-level key

So https://example.com/nested.claims.key gets treated as
https://example.com/nested -> claims -> key

### CODE REVIEW

**Comment by willthames** [comment]
I'll mark this as draft while I work on getting the CLA sorted.

File: test/jwt_test.cc [hunk 1]
```diff
@@ -520,27 +521,31 @@ TEST(JwtParseTest, GoodNestedJwt) {
    *      "key-3": true,
    *      "key-4": 9999
    *    }
-   *  }
+   *  },
```
Comment (nareddyt) [inline test/jwt_test.cc:524]: Can you add assertions to the test to verify these are accessible from the `Jwt` object?

Also suggest you add a few corner cases to this token, such as URL with escaped `/` characters, malformed URLs, etc.

File: src/struct_utils.cc [hunk 2]
```diff
@@ -105,8 +109,29 @@ StructUtils::FindResult StructUtils::GetStringList(
...
+
+  // Handle namespaced custom claims that by convention start with a URL - as
+  // most URLs contain dots, we extract the URL up to the FQDN and then split
+  // only the URL path by dots.
+  //
+  // e.g. for nested_names "https://example.com/claims.nested.key", name_vector
+  // would be {"https://example.com/claims", "nested", "key"}
+  if (absl::StartsWith(nested_names, "http://") ||
+      absl::StartsWith(nested_names, "https://")) {
+    const std::vector<std::string> url_components =
```
Comment (nareddyt) [inline src/struct_utils.cc:122]: I'm wary of doing this URL parsing manually with string manipulation instead of using an actual URL parser.

There's many things that can go wrong. At the very least, we should at least unescape the URL as `%2F` is the same as `/`. 

https://github.com/abseil/abseil-cpp/issues/158#issuecomment-1034695066

**Review by nareddyt — COMMENTED** [review]
> Private namespaced claims are recommended to be prefixed with URLs

Can you link to a specification that mentions this?

**Comment by willthames** [comment]
> > Private namespaced claims are recommended to be prefixed with URLs
> 
> Can you link to a specification that mentions this?

https://auth0.com/docs/secure/tokens/json-web-tokens/create-custom-claims#namespaced-guidelines

See also https://github.com/quarkusio/quarkus/issues/26185 for a similar issue in another library

It might only be auth0 who recommend this, but auth0 are a major advocate of JWT (they seem to be behind jwt.io for example)

File: test/jwt_test.cc (hunk 1 above)
Comment (willthames) [reply test/jwt_test.cc:524]: Haha, I went to a lot of effort to get the JWT token updated with the new claims and then somehow managed to not commit the tests that used it.

Added standard happy-path tests for now

**Comment by jgorinsky** [comment]
I ran into this issue while trying to extract namespaced JWT claims into headers with envoy. 

The JWT RFC (https://datatracker.ietf.org/doc/html/rfc7519#section-4.2) mentions using a domain name as a "Collision-Resistant Name" for claims and includes an example elsewhere using a domain namespaced claim:

```json
{
  "iss":"joe",
  "exp":1300819380,
  "http://example.com/is_root":true
}
```

**Comment by CelsoSantos** [comment]
I also ran into [this issue with Envoy](https://github.com/envoyproxy/envoy/issues/33603) and have been trying to figure out a way to handle this.
For Envoy it's not critical because we can use Lua scripting/filters to process the claim and place it in a header, but to solve this within Envoy source "natively" it would be way better if this lib had the support rather than having to work around it.

File: src/struct_utils.cc (hunk 2 above)
Comment (CelsoSantos) [reply src/struct_utils.cc:122]: While that's not wrong, I am thinking: Is there ever a situation where the contents of the token could potentially be URLencoded after having been b64-decoded from the `Authorization` header? Other than some user error maybe. And in that case, don't we want to fail because the token is not properly formatted?

Or is the assumption that the token is always obtained from said header wrong, for this use case, and we could actually run into this issue AND still want to process it as valid?

EDIT:
Actually, now that I'm revisiting this, I DO have one concern: What if the claim is a "url-like" string, but not starting with `http://` or `https://`, something like `example.org/some_claim` instead of `http://example.org/some_claim`? What then? Gracefully fail or let it try to unnest?
